Port to Python 3

ayushnawal opened the pull request.

ayushnawal committed:
Port to Python 3

ayushnawal commented:
@quozl please review!

quozl commented:
Thanks.  Reviewed.  Tested.  Please fix the warning about `GObject.source_remove`.

ayushnawal commented:
actually I tried, but I am not able to reproduce this warning, so I am not able to get and test any proper fix for it.

a little guidance can help me here.

Thanks :grin:

chimosky commented:
@ayushnawal try accessing all the buttons in the toolbar. This is the error.

```
/home/ibiam/Activities/iknowmyabcs/page.py:202: Warning: Source ID 1524 was not found when attempting to remove it
  GObject.source_remove(self.timeout)
```

ayushnawal committed:
Fix source_remove warning

ayushnawal commented:
@chimosky @quozl, please review

quozl commented:
Thanks.  Reviewed.  Waiting on https://github.com/sugarlabs/iknowmyabcs/pull/11 per https://github.com/sugarlabs/sugar-docs/blob/master/src/python-porting-guide.md#how-to-port-an-activity-to-python-3 "_making sure the activity works properly before porting, closing any solved issues, merging any pull requests or branches ..._".  You can progress this by adding those commits.

ayushnawal commented:
How can I proceed on #11 

Created account on http://activities.sugarlabs.org

I have very little idea on how to release an activity.

A little detailed guide will be much helpful for me to release this as well as many other activities.

quozl commented:
You can progress this pull request by adding those commits from the other pull request, i.e. by rebasing your branch on the other.

You should not attempt to release a Python 3 activity to http://activities.sugarlabs.org/ as we use that now for Python 2 activities.  We have no app store yet for Python 3 activities.

For detail on release, I can't guess what you don't know, so see the guide to contributing, and observe other activity releases, mostly by me, on the mailing list.  If you can't tell what I've done to release, then ask specific questions.

ayushnawal commented:
@quozl added commits from the other pull request as suggested.

quozl commented:
Thanks.  NEWS does not fully describe the nature of the changes since previous version.

ayushnawal commented:
Thanks, I didn't completely understand.

> NEWS does not fully describe the nature of the changes since the previous version.

by this are you saying to add the recent commits in the NEWS file or you meant to describe changes already mentioned in the file.

sorry, but I never worked on a NEWS file, want to learn.

quozl commented:
> > NEWS does not fully describe the nature of the changes since the previous version.
> 
> by this are you saying to add the recent commits in the NEWS file or you meant to describe changes already mentioned in the file.

By this I'm saying that the NEWS file is missing changes.

> sorry, but I never worked on a NEWS file, want to learn.

Look at the NEWS files of each of the Fructose activities.

ayushnawal committed:
update NEWS

ayushnawal commented:
I have added the recent commit of yours in `NEWS` as you suggested to add missing changes since the previous version.
@quozl, please review

quozl commented:
Thanks.  I've reviewed this pull request in detail again.

* the [aggregate change](https://github.com/sugarlabs/iknowmyabcs/pull/12/files) which at present consists of one line added to NEWS, changing `exec`, porting of the source files `page.py` and `utils/sprites.py`, and fixing the source_remove warning,
* each incremental commit, one by one.

I couldn't see why some of the incremental commits weren't in the aggregate change, until I used `gitk` to visualise your branch;

![gitk](https://user-images.githubusercontent.com/2936481/74205056-c1b19680-4cc9-11ea-918e-b1cc506fced6.png)

It looks like your branch diverged from master a long time ago.  I don't know what happened to cause that.  I suggest you pull your master branch from our master branch.  I suggest you rebase your port branch from a master branch.

ayushnawal commented:
rebased with the master branch as you suggested. 
it is following master now, with all commits as it should have been.

I think this happened because I rebased the branch with #11 as you suggested [here](https://github.com/sugarlabs/iknowmyabcs/pull/12#issuecomment-579552867) but those commits were already in the master branch.

quozl commented on utils/sprites.py:
```diff
@@ -251,7 +251,7 @@ def set_layer(self, layer=None):
         ''' Set the label drawn on the sprite '''
         self._extend_labels_array(i)
-        if type(new_label) is str or type(new_label) is unicode:
+        if type(new_label) is str or type(new_label) is str:
```
This line checks the same thing twice.  It is like `if False or False`, or `if True or True`.

quozl replied:
Also, other activities that contain sprites.py use isinstance instead of type; did you compare against them at all?

quozl reviewed (commented):
Thanks.  Reviewed.

I guess you focused on this repository only?  In my repositories I've found 20 different versions of sprites.py, each with different bugs.  This will be the 21st.

ayushnawal commented:
whoops .. I didn't notice that :sweat_smile: 

Fixed. Tested.

Although I didn't focus on `sprites.py` particularly in any other activity but can you please suggest me some of those, I will try to fix them.

Thanks

ayushnawal committed:
Fix sprites.py bug

quozl commented:
Next time, if you need to find `sprites.py`, try searching.

https://github.com/search?utf8=%E2%9C%93&q=org%3Asugarlabs+filename%3Asprites.py&type=Code&ref=advsearch&l=&l=

ayushnawal commented:
@srevinsaju this PR is ready from my side.
* Port to Python3
* Fix source_remove warning
* update news 
* Fix `sprites.py` bug 

No any errors and warnings

srevinsaju commented:
Thanks @ayushnawal, I will check it out when I have time, (my exams getting over on 20th)

ayushnawal commented:
@quozl  please [see](https://github.com/sugarlabs/iknowmyabcs/pull/12#issuecomment-586987327)

quozl commented:
Merged by hand.  https://github.com/sugarlabs/iknowmyabcs/commits/master
